In [1]:
from sklearn.metrics import mean_squared_error
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
housing = pd.read_csv('housing_processed.csv')
housing.head()

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0   1          60         65.0     8450            7            5       2003   
1   2          20         80.0     9600            6            8       1976   
2   3          60         68.0    11250            7            5       2001   
3   4          70         60.0     9550            7            5       1915   
4   5          60         84.0    14260            8            5       2000   

   YearRemodAdd  MasVnrArea ExterQual  ... GarageType_NA SaleType_COD  \
0          2003       196.0        Gd  ...             0            0   
1          1976         0.0        TA  ...             0            0   
2          2002       162.0        Gd  ...             0            0   
3          1970         0.0        TA  ...             0            0   
4          2000       350.0        Gd  ...             0            0   

  SaleType_CWD SaleType_Con SaleType_ConLD  SaleType_ConLI SaleType_ConLw  \
0            0            0              0               0              0   
1            0            0              0               0              0   
2            0            0              0               0              0   
3            0            0              0               0              0   
4            0            0              0               0              0   

   SaleType_New  SaleType_Oth  SaleType_WD  
0             0             0            1  
1             0             0            1  
2             0             0            1  
3             0             0            1  
4             0             0            1  

[5 rows x 238 columns]

In [3]:
housing_ml = housing.loc[:,housing.dtypes != object]

In [4]:
housing_ml.shape

(1460, 222)

In [5]:
# Split data as features and target
# take "SalePrice" values into its own 1-D array 
sale_price = housing_ml.pop('SalePrice')
type(sale_price)

pandas.core.series.Series

In [6]:
# pop removes the column
# "in place" operation
# now housing_ml has one less column
X = housing_ml
housing_ml.shape

(1460, 221)

In [7]:
y = sale_price.values
print(type(y), y.shape)

<class 'numpy.ndarray'> (1460,)


In [8]:
y.shape
# (1460,)
# it is equivalent to (1460)
# means it is a 1-d array

(1460,)

In [9]:
X = housing_ml.values

In [10]:
X.shape

(1460, 221)

In [11]:
X.shape[0] == y.shape[0]

True

In [12]:
# This uses Linear Regression
linear_model = LinearRegression()


After running the cells, add code to this notebook to achieve the following:
- Split the data for training and testing, to use 80 percent as training data.
use 21 as your randomization seed (so you achieve same results for us to grade).
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
random_state = 21
train_size = .8

In [13]:
#This uses the train test split from Scikit Learn
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=21)

In [14]:
print("X_train is ", X_train.shape, "x_test is ", X_test.shape)
print("y_train is ", y_train.shape, "y_test is ", y_test.shape)

X_train is  (1168, 221) x_test is  (292, 221)
y_train is  (1168,) y_test is  (292,)


In [15]:
linear_model.fit(X_train, y_train)

LinearRegression()

In [16]:
y_pred = linear_model.predict(X_train)
y_pred.shape

(1168,)

In [17]:
print("The accuracy of the training scores", linear_model.score(X_test, y_test)) 

The accuracy of the training scores 0.8127547098528062


In [18]:
print("The accuracy of the training rmse", np.sqrt(mean_squared_error(y_pred, y_train)))

The accuracy of the training rmse 20999.86660191205


In [19]:
#This is just used for the columns and maintained the id which was removed in the next line during the loop when it was passed
House_Columns = housing_ml.columns

In the following steps use training set to fit the model, and test set to evaluate it.
-  Analyze which feature alone would give the best prediction, list the scores and RMSE errors achieved by the top 10 predictors by score.
(create a loop to train a linear model for each feature and save the scores and RMSE and sort by score)

In [20]:
keys = []
score = []
rmse = []
for x in House_Columns:
  if x.startswith("Id"):
    pass
  else:
    X_House = housing_ml[[x]]
    linear_model.fit(X_House, y)
    y_pred = linear_model.predict(X_House)
    unsorted_score = linear_model.score(X_House, y)
    unsorted_rmse = np.sqrt(mean_squared_error(X_House, y)) 
    keys.append(x)
    score.append(unsorted_score)
    rmse.append(unsorted_rmse)
    House_score = dict(zip(keys, score))
    House_rmse = dict(zip(keys, rmse))

In [21]:
print("These are the scores in the dictionary")
House_score.items()

These are the scores in the dictionary


dict_items([('MSSubClass', 0.007103815434038263), ('LotFrontage', 0.11215612336205616), ('LotArea', 0.0696133153821119), ('OverallQual', 0.6256518924621182), ('OverallCond', 0.006061540238119112), ('YearBuilt', 0.27342162073249154), ('YearRemodAdd', 0.25715139084530325), ('MasVnrArea', 0.22336446466934434), ('BsmtFinSF1', 0.14932026665622322), ('BsmtFinSF2', 0.00012946164773608437), ('BsmtUnfSF', 0.04600128671622783), ('TotalBsmtSF', 0.37648109325168655), ('1stFlrSF', 0.36705686969596585), ('2ndFlrSF', 0.10197407763119004), ('LowQualFinSF', 0.0006556738936118744), ('GrLivArea', 0.5021486502718042), ('BsmtFullBath', 0.051584508782636984), ('BsmtHalfBath', 0.00028372553399325096), ('FullBath', 0.31434385485924576), ('HalfBath', 0.0807171713318704), ('BedroomAbvGr', 0.02829566527980476), ('KitchenAbvGr', 0.01847081344922341), ('TotRmsAbvGrd', 0.2848604068044378), ('Fireplaces', 0.21802253859013498), ('GarageCars', 0.41012393993308704), ('GarageArea', 0.388666759031819), ('WoodDeckSF', 0.1

In [22]:
print("These are the rmse in the dictionary")
House_rmse.items()

These are the rmse in the dictionary


dict_items([('MSSubClass', 197532.91111082333), ('LotFrontage', 197516.4756709386), ('LotArea', 187151.96770807984), ('OverallQual', 197577.54697060402), ('OverallCond', 197578.5011100235), ('YearBuilt', 195773.74364937475), ('YearRemodAdd', 195763.47202121938), ('MasVnrArea', 197454.89805812496), ('BsmtFinSF1', 197106.97903969162), ('BsmtFinSF2', 197541.75204371006), ('BsmtUnfSF', 197026.61125162937), ('TotalBsmtSF', 196507.54655618747), ('1stFlrSF', 196425.28696388486), ('2ndFlrSF', 197210.26552142468), ('LowQualFinSF', 197578.72606465206), ('GrLivArea', 196046.82618889908), ('BsmtFullBath', 197583.13462574067), ('BsmtHalfBath', 197583.5203857568), ('FullBath', 197582.01426202498), ('HalfBath', 197583.16345836694), ('BedroomAbvGr', 197580.89161085148), ('KitchenAbvGr', 197582.6251678779), ('TotRmsAbvGrd', 197577.25474771246), ('Fireplaces', 197582.88918949387), ('GarageCars', 197581.76105868), ('GarageArea', 197097.00438037736), ('WoodDeckSF', 197480.9725876627), ('OpenPorchSF', 1975

In [23]:
sorted_score = sorted(House_score.items(), key=lambda t: t[1], reverse=True) 
n = 10
print("These are the sorted top 10 highest scores ", sorted_score[:n])

These are the sorted top 10 highest scores  [('OverallQual', 0.6256518924621182), ('GrLivArea', 0.5021486502718042), ('ExterQual_Coded', 0.4659963342490334), ('KitchenQual_Coded', 0.4350717915853295), ('GarageCars', 0.41012393993308704), ('GarageArea', 0.388666759031819), ('TotalBsmtSF', 0.37648109325168655), ('1stFlrSF', 0.36705686969596585), ('BsmtQual_Coded', 0.34246746596334543), ('FullBath', 0.31434385485924576)]


In [24]:
sorted_rmse = sorted(House_rmse.items(), key=lambda t: t[1], reverse=True) 
print("These are the sorted top 10 highest rmse scores ", sorted_rmse[:n])

These are the sorted top 10 highest rmse scores  [('Electrical_Mix', 197583.5712198401), ('Heating_Floor', 197583.5712007741), ('Exterior1st_AsphShn', 197583.57110544422), ('Exterior1st_CBlock', 197583.57108811152), ('Exterior2nd_CBlock', 197583.57108811152), ('Condition2_RRAn', 197583.57097751155), ('RoofMatl_Roll', 197583.57097718221), ('Utilities_NoSeWa', 197583.57097544894), ('Exterior1st_BrkComm', 197583.57095985126), ('RoofMatl_ClyTile', 197583.57089745178)]


-  Select all possible 2 pairs of these top 10 predictors, and train 45 linear models, list the scores and RMSE errors achieved by the top 10 predictors by score.

In [25]:
#Make List
sorted_score10 = sorted_score[:n]
sorted_score1 = []
for i in range(10):
  sorted_score1.append(sorted_score[i][0])
sorted_score1

['OverallQual',
 'GrLivArea',
 'ExterQual_Coded',
 'KitchenQual_Coded',
 'GarageCars',
 'GarageArea',
 'TotalBsmtSF',
 '1stFlrSF',
 'BsmtQual_Coded',
 'FullBath']

In [26]:
from itertools import combinations
comb = combinations(sorted_score1, 2)
combination_result =[]
x_comb = []
y_comb = []
for comblist in list(comb):
  combination_result.append(comblist)

print(len(combination_result)) 
print(combination_result)

45
[('OverallQual', 'GrLivArea'), ('OverallQual', 'ExterQual_Coded'), ('OverallQual', 'KitchenQual_Coded'), ('OverallQual', 'GarageCars'), ('OverallQual', 'GarageArea'), ('OverallQual', 'TotalBsmtSF'), ('OverallQual', '1stFlrSF'), ('OverallQual', 'BsmtQual_Coded'), ('OverallQual', 'FullBath'), ('GrLivArea', 'ExterQual_Coded'), ('GrLivArea', 'KitchenQual_Coded'), ('GrLivArea', 'GarageCars'), ('GrLivArea', 'GarageArea'), ('GrLivArea', 'TotalBsmtSF'), ('GrLivArea', '1stFlrSF'), ('GrLivArea', 'BsmtQual_Coded'), ('GrLivArea', 'FullBath'), ('ExterQual_Coded', 'KitchenQual_Coded'), ('ExterQual_Coded', 'GarageCars'), ('ExterQual_Coded', 'GarageArea'), ('ExterQual_Coded', 'TotalBsmtSF'), ('ExterQual_Coded', '1stFlrSF'), ('ExterQual_Coded', 'BsmtQual_Coded'), ('ExterQual_Coded', 'FullBath'), ('KitchenQual_Coded', 'GarageCars'), ('KitchenQual_Coded', 'GarageArea'), ('KitchenQual_Coded', 'TotalBsmtSF'), ('KitchenQual_Coded', '1stFlrSF'), ('KitchenQual_Coded', 'BsmtQual_Coded'), ('KitchenQual_Coded

In [27]:
for x in range(45):
  x_comb = combination_result[x][0]
  y_comb = combination_result[x][1]
  X_House = housing_ml[[x_comb]]
  y_house = housing_ml[[y_comb]]
  linear_model.fit(X_House, y_house)
  y_pred = linear_model.predict(X_House)
  unsorted_score_comb = linear_model.score(X_House, y_house)
  unsorted_rmse_comb = np.sqrt(mean_squared_error(X_House, y_house))
  print(combination_result[x][0], " and ", combination_result[x][1], "score is ", unsorted_score_comb," rmse is ", unsorted_rmse_comb)

OverallQual  and  GrLivArea score is  0.3516578120691849  rmse is  1597.8929645090116
OverallQual  and  ExterQual_Coded score is  0.5274804461466409  rmse is  3.8475218852573465
OverallQual  and  KitchenQual_Coded score is  0.4533743408949561  rmse is  3.7400095230997357
OverallQual  and  GarageCars score is  0.3608053097696037  rmse is  4.471753050255096
OverallQual  and  GarageArea score is  0.3158684549059215  rmse is  513.1556013850737
OverallQual  and  TotalBsmtSF score is  0.2892379811797141  rmse is  1138.8482544111582
OverallQual  and  1stFlrSF score is  0.2267891353818806  rmse is  1219.178564129503
OverallQual  and  BsmtQual_Coded score is  0.396118069597755  rmse is  2.822730631707192
OverallQual  and  FullBath score is  0.3031600399564057  rmse is  4.683547420316801
GrLivArea  and  ExterQual_Coded score is  0.1900838435071074  rmse is  1601.5779930456772
GrLivArea  and  KitchenQual_Coded score is  0.17687302400791638  rmse is  1601.4592365406525
GrLivArea  and  GarageCars s

-  Train a single model using all features. Calculate RMSE and score. Observe how much of the prediction power was in the 2 pairs, vs all features.

In [28]:
linear_model.fit(X, y)
y_pred6 = linear_model.predict(X)
print("The accuracy of the score using all the data: ", linear_model.score(X, y))
print("The accuracy of the rmse using al the data: ", np.sqrt(mean_squared_error(y_pred6, y)))

The accuracy of the score using all the data:  0.9184728890914109
The accuracy of the rmse using al the data:  22675.41074544839



- As of the scores for all features is high percentage 91.85% and rmse for all feature is also high. It has a better accuracy<br>
- As for the top10 with 2 models, the scores and rmse is unpredictable.

- Use the 5NN and 10NN regressor with all features, and list the RMSE and score for these 2 models 
https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html
observe if the results are better than linear regression? 


In [29]:
from sklearn.neighbors import KNeighborsRegressor
neigh5 = KNeighborsRegressor(n_neighbors=5)
neigh10 = KNeighborsRegressor(n_neighbors=10)

In [30]:
neigh5.fit(y_pred, y)
print("Using 5NN, this predict for the Full Bath is ", neigh5.predict([[1]]))
neigh10.fit(y_pred, y)
print("Using 10NN, this predict for the Overall Quality is ", neigh10.predict([[1]]))

Using 5NN, this predict for the Full Bath is  [96600.]
Using 10NN, this predict for the Overall Quality is  [99690.]


In [31]:
for x in range(45):
  x_comb = combination_result[x][0]
  y_comb = combination_result[x][1]
  X_House = housing_ml[[x_comb]]
  y_house = housing_ml[[y_comb]]
  neigh5.fit(X_House, y_house)
  neigh10.fit(X_House, y_house)
  print("Using 5NN, this predict for ", combination_result[x][0], " and ", combination_result[x][1], " is ", neigh5.predict([[1]]))

Using 5NN, this predict for  OverallQual  and  GrLivArea  is  [[591.2]]
Using 5NN, this predict for  OverallQual  and  ExterQual_Coded  is  [[1.4]]
Using 5NN, this predict for  OverallQual  and  KitchenQual_Coded  is  [[1.2]]
Using 5NN, this predict for  OverallQual  and  GarageCars  is  [[0.4]]
Using 5NN, this predict for  OverallQual  and  GarageArea  is  [[110.8]]
Using 5NN, this predict for  OverallQual  and  TotalBsmtSF  is  [[343.4]]
Using 5NN, this predict for  OverallQual  and  1stFlrSF  is  [[591.2]]
Using 5NN, this predict for  OverallQual  and  BsmtQual_Coded  is  [[2.2]]
Using 5NN, this predict for  OverallQual  and  FullBath  is  [[0.6]]
Using 5NN, this predict for  GrLivArea  and  ExterQual_Coded  is  [[1.8]]
Using 5NN, this predict for  GrLivArea  and  KitchenQual_Coded  is  [[1.4]]
Using 5NN, this predict for  GrLivArea  and  GarageCars  is  [[0.6]]
Using 5NN, this predict for  GrLivArea  and  GarageArea  is  [[158.8]]
Using 5NN, this predict for  GrLivArea  and  TotalB

In [32]:
 for x in range(45):
  x_comb = combination_result[x][0]
  y_comb = combination_result[x][1]
  X_House = housing_ml[[x_comb]]
  y_house = housing_ml[[y_comb]]
  neigh10.fit(X_House, y_house)
  print("Using 10NN, this predict for ", combination_result[x][0], " and ", combination_result[x][1], " is ", neigh10.predict([[1]]))

Using 10NN, this predict for  OverallQual  and  GrLivArea  is  [[873.2]]
Using 10NN, this predict for  OverallQual  and  ExterQual_Coded  is  [[1.6]]
Using 10NN, this predict for  OverallQual  and  KitchenQual_Coded  is  [[1.5]]
Using 10NN, this predict for  OverallQual  and  GarageCars  is  [[0.6]]
Using 10NN, this predict for  OverallQual  and  GarageArea  is  [[148.2]]
Using 10NN, this predict for  OverallQual  and  TotalBsmtSF  is  [[540.5]]
Using 10NN, this predict for  OverallQual  and  1stFlrSF  is  [[771.5]]
Using 10NN, this predict for  OverallQual  and  BsmtQual_Coded  is  [[2.3]]
Using 10NN, this predict for  OverallQual  and  FullBath  is  [[0.9]]
Using 10NN, this predict for  GrLivArea  and  ExterQual_Coded  is  [[1.9]]
Using 10NN, this predict for  GrLivArea  and  KitchenQual_Coded  is  [[2.]]
Using 10NN, this predict for  GrLivArea  and  GarageCars  is  [[0.5]]
Using 10NN, this predict for  GrLivArea  and  GarageArea  is  [[128.5]]
Using 10NN, this predict for  GrLivArea

- Which regressor is better for inference?
 - Comparing with the 2 Models from Top 10 and the all features for 5NN and 10 NN. Its prediction from 5NN to 10NN gets to bigger number. However, the better predictor is for the all features and with a higher NN such as 10NN
 - Using the score and rmse would be a better regressor for inference than using KNN because scores has a range on or within 1. Rmse has a range closer to mean. 